src/snowcli/cli/nativeapp/artifacts.py

```diff
@@ -0,0 +1,175 @@
+    """
+
+    dest_path: Path
+
+    def __init__(self, dest_path: Path):
+        super().__init__()
+        self.dest_path = dest_path
+
+    def __str__(self):
+        return self.__doc__ + f"\ndest_path = {self.dest_path}"
+
+
+class OutsideDeployRootError(ArtifactError):
+    """
+    The specified path is outside of the deploy root.
+    This can happen when a relative path with ".." is provided.
+    """
+
+    dest_path: Path
+    deploy_root: Path
+
+    def __init__(self, dest_path: Path, deploy_root: Path):
+        super().__init__()
+        self.dest_path = dest_path
+        self.deploy_root = deploy_root
+
+    def __str__(self):
+        return (
+            self.__doc__
+            + f"\ndest_path = {self.dest_path}"
+            + f"\ndeploy_root = {self.deploy_root}"
+        )
+
+
+@dataclass
+class SrcDestPair:
+    """
+    A pair of paths, relative to some context.
+    """
+
+    src: str
+    dest: str
+
+
+def is_glob(s: str) -> bool:
+    return "*" in s
+
+
+def specifies_directory(s: str) -> bool:
+    """
+    Does the path (as seen from the project definition) refer to
+    a directory? For destination paths, we enforce the usage of a
+    trailing slash (i.e. \ for windows; / for others).
+
+    This means that to put a file in the root of the stage, we need
+    to specify "./" as its destination, or omit it (but only if the
+    file already lives in the project root).
+    """
+    return s.endswith(os.sep)
+
+
+def symlink_or_copy(src: Path, dst: Path, makedirs=True) -> None:
+    """
+    Tries to create a symlink to src at dst; failing that (i.e. in Windows
+    without Administrator / Developer Mode) copies the file from src to dst instead.
+    If makedirs is True, the directory hierarchy above dst is created if any
+    of those directories do not exist.
+    """
+    if makedirs:
+        dst.parent.mkdir(parents=True, exist_ok=True)
+    try:
+        os.symlink(src, dst)
+    except OSError:
+        shutil.copy(src, dst)
+
+
+def translate_artifact(item: Union[dict, str]) -> SrcDestPair:
+    if isinstance(item, dict):
+        if "dest" in item:
+            return SrcDestPair(item["src"], item["dest"])
+        else:
+            return SrcDestPair(item["src"], item["src"])
+
+    elif isinstance(item, str):
+        return SrcDestPair(item, item)
+
+    else:
+        # XXX: validation should have caught this
+        raise ArtifactError("Item is not a valid artifact!")
+
+
+def build_bundle(project_root: Path, deploy_root: Path, artifacts: List[SrcDestPair]):
+    """
+    Prepares a local folder (deploy_root) with configured app artifacts.
+    This folder can then be uploaded to a stage.
+    """
+    if deploy_root.exists() and not deploy_root.is_dir():
+        raise ValueError(f"Deploy root {deploy_root} exists, but is not a directory!")
+
+    for artifact in artifacts:
+        # build the list of source files
+        source_paths: List[Path]
```

Comment:
What do you think about encapsulating this logic into a function? It's a quite long logic in this for loop